Added VoidService and gateway#void

abhishekgupta5 opened the pull request:
**Issue:** https://linear.app/nebulab-retainers/issue/ABU-10/implement-soliduspaytomorrowgatewayvoid

**Brief:**
- [x] Added VoidService in payTomorrow Client
- [x] Added specs for ^
- [x] Use the service in `gateway.rb`
- [x] Update gateway specs

Naokimi commented on app/services/solidus_pay_tomorrow/client/void_service.rb:
```diff
@@ -0,0 +1,29 @@
+    class VoidService < SolidusPayTomorrow::Client::BaseService
+      attr_reader :order_token
+
+      CANCEL_ENDPOINT = 'api/ecommerce/application/:order_token/cancel'
```
doesn't it need `.freeze` at the end?

abhishekgupta5 replied:
`frozen_string_literal: true` at the beginning of the file tells Ruby that all strings in the file are implicitly frozen.
Hence, in this case, freeze isn't needed for string types

Naokimi commented on spec/services/solidus_pay_tomorrow/client/void_service_spec.rb:
```diff
@@ -0,0 +1,17 @@
+
+  describe '#call' do
+    it 'successful cancellation', vcr: { cassette_name: 'client/void/success' } do
+      response = described_class.call(order_token: 'order_token')
```
the service successfully canceled an order with a token `'order_token'`? 🤯

abhishekgupta5 replied:
No. The test passed with a valid token, but need not have it in version control, hence replaced a valid_token with `order_token` in yml and test.

Is this not the recommended way?

abhishekgupta5 replied:
Ignore this^ comment. We decided not to use VCR with this extension

abhishekgupta5 committed:
Added VoidService

* Client code for implementing void or cancellation
* Specs for above^
* Note: An order can be cancelled when user has entered their
        details on PayTomorrow URL and the state is - S: Ready for Settle
        (An uncaptured payment)
        If a payment is captured, void will err. It should be
        credit/refund in this case
        Ref: https://docs.paytomorrow.com/docs/Integration/SAI/api-stages
             for all the stages

abhishekgupta5 committed:
Use VoidService in gateway

* Both success and failure should return an
  ActiveMerchant::Billing::Response
* Specs^

DanielePalombo commented on spec/models/solidus_pay_tomorrow/gateway_spec.rb:
```diff
@@ -41,15 +41,38 @@
+        lender: nil }.stringify_keys!
+    end
+
+    it 'returns an active merchant billing success response' do
+      allow(SolidusPayTomorrow::Client::VoidService).to receive(:call).with(
+        order_token: payment.response_code,
+        payment_method: payment_method
+      ).and_return(void_success_response)
+      result = described_class.new.void(payment.response_code,
+        originator: payment)
+      expect(result).to be_an_instance_of(ActiveMerchant::Billing::Response)
+      expect(result).to be_success
+    end
+
+    it 'returns an active merchant billing failure response' do
+      allow(SolidusPayTomorrow::Client::VoidService).to receive(:call).and_raise(StandardError)
+      result = described_class.new.void(payment.response_code,
+        originator: payment)
+      expect(result).to be_an_instance_of(ActiveMerchant::Billing::Response)
+      expect(result).not_to be_success
```
The 2 examples looks very strange.
Do you mind creating a different context for them?

abhishekgupta5 replied:
Added [here](https://github.com/nebulab/solidus_pay_tomorrow/pull/5/commits/ce1e367348f5d6f849f557f6c51a94ec6c50478e#diff-29658277569ec4cc1cf064679c583c4ee95610938fab468405b20eab45d82cb0)

DanielePalombo reviewed (approved):
LGTM
